### PR TITLE
Set CDI's application TCCL during shutdown events

### LINKED ISSUE #29277: CDI does not set the TCCL during shutdown
**Describe the bug**  
CDI creates its own TCCL and sets it before firing CDI observers on application startup, it does not do so on application shutdown.

This issue was reported by a customer that experienced the following because there was a different TCCL during startup and shutdown

```[7/11/24, 9:29:37:811 CEST] 0000073e Bootstrap    E org.jboss.weld.bootstrap.events.BeforeShutdownImpl fire 

                                javax.enterprise.event.ObserverException

   at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)

   at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)

   at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)

   at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)

   at java.base/java.lang.reflect.ReflectAccess.newInstance(ReflectAccess.java:128)

   at java.base/jdk.internal.reflect.ReflectionFactory.newInstance(ReflectionFactory.java:347)

   at java.base/java.lang.Class.newInstance(Class.java:645)

   at org.jboss.weld.security.NewInstanceAction.run(NewInstanceAction.java:33)

   at java.base/java.security.AccessController.doPrivileged(AccessController.java:569)

   at org.jboss.weld.injection.Exceptions.rethrowException(Exceptions.java:40)

   at org.jboss.weld.injection.Exceptions.rethrowException(Exceptions.java:78)

   at org.jboss.weld.injection.StaticMethodInjectionPoint.invoke(StaticMethodInjectionPoint.java:103)

   at org.jboss.weld.injection.StaticMethodInjectionPoint.invoke(StaticMethodInjectionPoint.java:85)

   at org.jboss.weld.injection.MethodInvocationStrategy$SimpleMethodInvocationStrategy.invoke(MethodInvocationStrategy.java:168)

   at org.jboss.weld.event.ObserverMethodImpl.sendEvent(ObserverMethodImpl.java:330)

   at org.jboss.weld.event.ExtensionObserverMethodImpl.sendEvent(ExtensionObserverMethodImpl.java:123)

   at org.jboss.weld.event.ObserverMethodImpl.sendEvent(ObserverMethodImpl.java:308)

   at org.jboss.weld.event.ObserverMethodImpl.notify(ObserverMethodImpl.java:286)

   at javax.enterprise.inject.spi.ObserverMethod.notify(ObserverMethod.java:124)

   at org.jboss.weld.util.Observers.notify(Observers.java:166)

   at org.jboss.weld.event.ObserverNotifier.notifySyncObservers(ObserverNotifier.java:285)

   at org.jboss.weld.event.ObserverNotifier.notify(ObserverNotifier.java:273)

   at org.jboss.weld.event.ObserverNotifier.fireEvent(ObserverNotifier.java:177)

   at org.jboss.weld.event.ObserverNotifier.fireEvent(ObserverNotifier.java:171)

   at org.jboss.weld.bootstrap.events.AbstractContainerEvent.fire(AbstractContainerEvent.java:53)

   at org.jboss.weld.bootstrap.events.BeforeShutdownImpl.fire(BeforeShutdownImpl.java:41)

   at org.jboss.weld.bootstrap.events.BeforeShutdownImpl.fire(BeforeShutdownImpl.java:32)

   at org.jboss.weld.bootstrap.WeldRuntime.shutdown(WeldRuntime.java:69)

   at org.jboss.weld.bootstrap.WeldBootstrap.shutdown(WeldBootstrap.java:116)

   at com.ibm.ws.cdi.impl.weld.WebSphereCDIDeploymentImpl$1.run(WebSphereCDIDeploymentImpl.java:683)

   at com.ibm.ws.cdi.impl.weld.WebSphereCDIDeploymentImpl$1.run(WebSphereCDIDeploymentImpl.java:680)

   at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)

   at com.ibm.ws.cdi.impl.weld.WebSphereCDIDeploymentImpl.shutdown(WebSphereCDIDeploymentImpl.java:680)

   at com.ibm.ws.cdi.impl.CDIContainerImpl.applicationStopped(CDIContainerImpl.java:262)

   at com.ibm.ws.cdi.liberty.CDIRuntimeImpl.applicationStopped(CDIRuntimeImpl.java:502)

   at com.ibm.ws.container.service.state.internal.ApplicationStateManager.fireStopped(ApplicationStateManager.java:162)

   at com.ibm.ws.container.service.state.internal.StateChangeServiceImpl.fireApplicationStopped(StateChangeServiceImpl.java:67)

   at com.ibm.ws.app.manager.module.internal.SimpleDeployedAppInfoBase.uninstallApp(SimpleDeployedAppInfoBase.java:683)

   at com.ibm.ws.app.manager.module.internal.DeployedAppInfoBase.uninstallApp(DeployedAppInfoBase.java:230)

   at com.ibm.ws.app.manager.ear.internal.EARApplicationHandlerImpl.uninstall(EARApplicationHandlerImpl.java:100)

   at com.ibm.ws.app.manager.internal.statemachine.StopAction.execute(StopAction.java:85)

   at com.ibm.ws.app.manager.internal.statemachine.ApplicationStateMachineImpl.enterState(ApplicationStateMachineImpl.java:1394)

   at 

   ... 35 more```

### PR DESCRIPTION
**Pull request by benjamin-confino**
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

First checkbox is a TODO, this needs further socalizatoin since it changes behaviour. See this previous brief discussion: https://ibm-cloud.slack.com/archives/C6K9D2D6J/p1722627570247169?thread_ts=1722351151.644479&cid=C6K9D2D6J

resolves #29277

### CODE REVIEW

**Comment by benjamin-confino** [comment]
#build

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java [hunk 1]
```diff
@@ -41,7 +42,6 @@
 import com.ibm.websphere.ras.TraceComponent;
 import com.ibm.ws.cdi.CDIException;
 import com.ibm.ws.cdi.CDIService;
-import com.ibm.ws.cdi.extension.WebSphereCDIExtension;
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java:44]: Why has this import been removed and the fully-qualified version used everywhere?

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java [hunk 2]
```diff
@@ -215,39 +215,33 @@ public void endInitialization(WebSphereCDIDeployment webSphereCDIDeployment) thr
...
+
+    private void fireEventWithContext(Consumer<BeanDeploymentModule> event, Application application, CDIContainerEventManager eventManager) throws CDIException {
+        WebSphereCDIDeployment deployment = getDeployment(application);
+        if (deployment != null) {
+            BeanDeploymentModules modules = deployment.getServices().get(BeanDeploymentModules.class);
+            if (modules != null) {
+                for (BeanDeploymentModule module : modules) {
+                    if (!module.isWebModule()) {
+                        String id = module.getId();
+                        WebSphereBeanDeploymentArchive bda = deployment.getBeanDeploymentArchive(id);
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java:246]: I don't like this way of getting a BDA for a module, it seems very breakable and doesn't match the way Weld's `BeanDeploymentModules` navigates from a `BeanManagerImpl` to a `BeanDeploymentModule`.

Probably warrants a tech debt item.

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java [hunk 3]
```diff
@@ -467,8 +467,8 @@ public void applicationStarting(ApplicationInfo appInfo) throws StateChangeExcep
 
             ClassLoader appCL = getRealAppClassLoader(application);
             if (appCL != null) {
-                ClassLoader newCL = classLoadingSRRef.getServiceWithException().createThreadContextClassLoader(appCL);
-                application.setTCCL(newCL);
+                ClassLoader newCDISpecificCL = classLoadingSRRef.getServiceWithException().createThreadContextClassLoader(appCL);
+                application.setTCCL(newCDISpecificCL);
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java]: I'm unclear about why this change is made, what's CDISpecific about this classloader?

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java [hunk 4]
```diff
@@ -215,39 +215,33 @@ public void endInitialization(WebSphereCDIDeployment webSphereCDIDeployment) thr
...
-                BeanDeploymentModules modules = deployment.getServices().get(BeanDeploymentModules.class);
-                if (modules != null) {
-                    for (BeanDeploymentModule module : modules) {
-                        if (!module.isWebModule()) {
-                            eventManager.fireShutdownEvent(module);
+            fireEventWithContext(eventManager::fireShutdownEvent, application, eventManager);
+        }
+    }
+
+    private void fireEventWithContext(Consumer<BeanDeploymentModule> event, Application application, CDIContainerEventManager eventManager) throws CDIException {
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java]: 1. Method needs Javadoc
2. `eventManager` parameter is not used and should be removed.
3. Nothing about the code in the method involves firing events. Suggest renaming to `runForEachModuleWithContext` and renaming the `event` parameter to `action`.

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java [hunk 5]
```diff
@@ -639,16 +635,17 @@ public String getCurrentApplicationContextID() {
      * @return
      * @throws CDIException
      */
+    @SuppressWarnings("deprecation") //Until nobody uses this deprecated interface we still need to read it
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java]: Not clear which deprecated interface is being referred to.

**Review by Azquelt — CHANGES_REQUESTED** [review]
FAT tests to ensure the TCCL and app context are set correctly within `Startup`, `Shutdown` and `BeforeShutdown` events need to be added, though the code does _look_ correct to me.

Also see the comments above.

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java (hunk 1 above)
Comment (benjamin-confino) [reply dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java:44]: The only way to do a `@SuppressWarnings("deprecation")` for an import statement is to put it on the class itself. Which is far to broad and might in the future hide a deprecation we actually need to fix. 

The workaround is to use the fully-qualified version inside methods so you can suppress the warning there.

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java (hunk 5 above)
Comment (Azquelt) [reply dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java]: In the IDE, it's more obvious because `WebSphereCDIExtension` is struck through.

The comment should either mention `WebSphereCDIExtension` or be removed.

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java (hunk 3 above)
Comment (benjamin-confino) [reply dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java]: This classloader is created on line 470 and is only ever set onto the thread context when CDI methods are running.

I made the change because David Zavala thought for a moment that CDI was destroying the main application classloader, this should prevent anyone else being confused about that.

Comment (Azquelt) [reply dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java]: Unfortunately, I don't think that's how it actually works. Although the method is called `createThreadContextClassLoader` it actually returns the same TCCL if passed the same app classloader, and increments a reference count (see [`ClassLoadingServiceImpl`](https://github.com/OpenLiberty/open-liberty/blob/c36e2355109d7838e558d2da9fb78b615b6dc35a/dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ClassLoadingServiceImpl.java#L650)).

Having the methods named `create` and `destroy` is misleading, but the TCCL we're getting back is not CDI specific.

That's actually important since libraries may use the TCCL as a key to a map (as I think MP Config does).

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java (hunk 2 above)
Comment (benjamin-confino) [reply dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java:246]: I opened https://github.com/OpenLiberty/open-liberty/issues/29279

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java (hunk 3 above)
Comment (benjamin-confino) [reply dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java]: It looks like you're correct, but the old name still shocked and confused. I'll ask David what he suggests we call it.

Comment (Azquelt) [reply dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java]: Ideally we would change `ClassLoadingService` to have methods named `acquire` and `release` and document their effects but it's SPI so we can't remove the existing methods. We could maybe improve the Javadoc.

I'm not sure we can improve anything by changing the variable name.

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java [hunk 6]
```diff
@@ -215,39 +215,41 @@ public void endInitialization(WebSphereCDIDeployment webSphereCDIDeployment) thr
...
+
+    /**
+     * This method loops through each module (that is not a WebModule) in an application. For each one it sets
+     * the component metadata associated with that module as well as the TCCL associated with the application
+     * then it runs an action. The action is likely an event form CDIContainerEventManager
+     *
+     * @param action The action to run
+     * @param application the application to run actions
+     */
+    private void runForEachModuleWithContext(Consumer<BeanDeploymentModule> action, Application application) throws CDIException {
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java]: I'd forgotten that it only runs for non-web modules.

I would suggest renaming to `runForEachNonWebModuleWithContext`, or restructuring the code into more easily-named parts.

We should also add a comment to indicate why we want to only run for non-web modules in the existing locations that this is called so that it's easier to track down any problems in this area in future. I'm guessing that the integration with the web container calls these events for web modules for some reason, but we'd need to check.

Comment (benjamin-confino) [reply dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java]: Since this is a private method and there's no likelihood we'll need to do something similar for a different combination of modules anytime soon, I think renaming is the way to go.

Regarding the comment. `!module.isWebModule()` originates in [this commit](https://github.com/OpenLiberty/open-liberty/commit/5178c4fc788f6a63aaae7956af53873db9c28531), so I'll ask Tom if he remembers why he did it this way.

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java (hunk 3 above)
Comment (benjamin-confino) [reply dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java]: I spoke to Dave and he recommended `appTCCL`

**Comment by benjamin-confino** [comment]
#build

File: dev/io.openliberty.cdi.4.0.internal_fat/fat/src/io/openliberty/cdi40/internal/fat/startupEvents/sharedLib/AbstractObserver.java [hunk 7]
```diff
@@ -149,6 +159,28 @@ private void runAllJNDITests() throws Exception {
...
+        if (tccl.getClass().getName().contains("ThreadContextClassLoader")) {
+            Method getID = tccl.getClass().getMethod("getKey");
+            id = (String) getID.invoke(tccl);
+        } else if (tccl.getClass().getName().contains("ThreadContextClassLoader")) {
+            Method getID = tccl.getClass().getMethod("getKey");
+            Object classLoaderIdentity = getID.invoke(tccl);
+            id = classLoaderIdentity.toString();
+        } else {
+            Assert.fail("TCCL was neither a ThreadContextClassLoader or an AppClassLoader");
+        }
```
Comment (Azquelt) [inline dev/io.openliberty.cdi.4.0.internal_fat/fat/src/io/openliberty/cdi40/internal/fat/startupEvents/sharedLib/AbstractObserver.java]: 1. both `if` conditions appear to check for the same thing
3. The assert message suggests that either `ThreadContextClassLoader` or `AppClassLoader` is acceptable - shouldn't it always be `ThreadContextClassLoader`?
2. I don't see what logic from `ClassLoadingServiceImpl` this code is duplicating
3. If we know that the classloader is a `ThreadContextClassLoader`, would it be sufficient to ensure it can load a resource from the application to check that we have the right one?

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java (hunk 1 above)
Comment (Azquelt) [reply dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java:44]: Ah, ok. I'm not sure we want to adopt that as a general strategy, but it's ok here.

File: dev/io.openliberty.cdi.4.0.internal_fat/fat/src/io/openliberty/cdi40/internal/fat/startupEvents/sharedLib/AbstractObserver.java (hunk 7 above)
Comment (benjamin-confino) [reply dev/io.openliberty.cdi.4.0.internal_fat/fat/src/io/openliberty/cdi40/internal/fat/startupEvents/sharedLib/AbstractObserver.java]: 1 & 2. My mistake. The second if should be for AppCllassLoader
3. The method `getClassLoaderIdentifier`
4. I'll ask the class loading people to make sure there's no possibility of getting a false positive that way.

Comment (Azquelt) [reply dev/io.openliberty.cdi.4.0.internal_fat/fat/src/io/openliberty/cdi40/internal/fat/startupEvents/sharedLib/AbstractObserver.java]: 3. aha, I see it, thank you

**Review by Azquelt — DISMISSED** [review]

**Comment by benjamin-confino** [comment]
The previous build is still visible and looking great, but starting another just to be safe since its two weeks old

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

File: dev/io.openliberty.cdi.4.0.internal_fat/fat/src/io/openliberty/cdi40/internal/fat/startupEvents/sharedLib/TestLoggingObserver.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            //Omitting a stack because all these test classes already log out their stack
+            //(The nature of these tests means that if one fails you want the other stacks for
+            //comparason)
+            System.out.println("===TEST FAIL " + name + " " + e);
+        } catch (Exception e) {
+            System.out.println("===TEST EXCEPTION " + name + " " + e);
+        } catch (Throwable e) {
+            System.out.println("===TEST EXCEPTION " + name + " " + e);
+            throw e;
+        }
```
Comment (Azquelt) [inline dev/io.openliberty.cdi.4.0.internal_fat/fat/src/io/openliberty/cdi40/internal/fat/startupEvents/sharedLib/TestLoggingObserver.java:42]: Did you mean to log the stack for the other cases, or are they ok too?

Comment (benjamin-confino) [reply dev/io.openliberty.cdi.4.0.internal_fat/fat/src/io/openliberty/cdi40/internal/fat/startupEvents/sharedLib/TestLoggingObserver.java:42]: That comment applies to all of them. `+ e` only printed out the message but not the stack.

File: dev/io.openliberty.cdi.4.0.internal_fat/fat/src/io/openliberty/cdi40/internal/fat/startupEvents/sharedLib/TestLoggingObserver.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+@LoggedTest
+@Interceptor
+@Priority(1)
+public class TestLoggingObserver {
+
+    @AroundInvoke
+    public Object runlogTestMethod(InvocationContext invocationContext) {
+
+        String name = "" + invocationContext.getTarget().getClass().getName().replaceAll("\\$Proxy\\$_\\$\\$_WeldSubclass", "") + "." + invocationContext.getMethod().getName();
+        Object returnValue = null;
```
Comment (Azquelt) [inline dev/io.openliberty.cdi.4.0.internal_fat/fat/src/io/openliberty/cdi40/internal/fat/startupEvents/sharedLib/TestLoggingObserver.java:26]: Have you run this in full mode - to ensure that this pattern replacement works on all CDI versions?

Comment (Azquelt) [reply dev/io.openliberty.cdi.4.0.internal_fat/fat/src/io/openliberty/cdi40/internal/fat/startupEvents/sharedLib/TestLoggingObserver.java:26]: 1. This is a CDI 4+ test, so there aren't that many repeats.
2. Yes it ran in full mode in the build.

**Review by Azquelt — APPROVED** [review]
Looks good, thanks.

**Comment by benjamin-confino** [comment]
#libby
